### ISSUE TITLE
TensorFlow Lite error on iOS: "Make sure you apply/link the Flex delegate before inference."

### ISSUE THREAD

**Issue by alexdmiller** [tensorflow/tensorflow#44997]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): macOS 10.15, iPhone Simulator 12
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: iPhone Simulator 12
- TensorFlow installed from (source or binary): Installed from Cocoa Pod using the following specification:

  ```
  pod 'TensorFlowLiteSwift', '0.0.1-nightly.20200916', :subspecs => ['CoreML', 'Metal'] 
  pod 'TensorFlowLiteSelectTfOps', '0.0.1-nightly.20200916'
  ```

**Describe the current behavior**

When I load my model using TensorFlow Lite and attempt to invoke it (`interpreter.invoke()`), I get the following error:

```
2020-11-18 16:08:54.145490-0800 OpenRDTCV[13328:3534503] Initialized TensorFlow Lite runtime.
2020-11-18 16:08:56.458094-0800 OpenRDTCV[13328:3534503] Regular TensorFlow ops are not supported by this interpreter. Make sure you apply/link the Flex delegate before inference.
2020-11-18 16:08:56.458232-0800 OpenRDTCV[13328:3534503] Node number 312 (FlexSize) failed to prepare.
```

I have read and followed [Select TensorFlow operators](https://www.tensorflow.org/lite/guide/ops_select) and added the `-force_load` flag (see screenshot below).

![Screen Shot 2020-11-18 at 4 19 39 PM](https://user-images.githubusercontent.com/926408/99604269-eb66bf00-29b9-11eb-9d51-5664451759a6.png)

I have also tried a variety of versions of TensorFlowLiteSwift/TensorFlowLiteSelectTfOps. However, the error persists. Any help appreciated!

Here is the relevant Swift source up until the failing `invoke()` line:

```swift
    guard let pixelBuffer = CVPixelBuffer.buffer(from: image) else {
      print("Could not convert image to CV buffer")
      return []
    }
    
    let sourcePixelFormat = CVPixelBufferGetPixelFormatType(pixelBuffer)
    assert(sourcePixelFormat == kCVPixelFormatType_32ARGB ||
             sourcePixelFormat == kCVPixelFormatType_32BGRA ||
               sourcePixelFormat == kCVPixelFormatType_32RGBA)

    let imageChannels = 4
    assert(imageChannels >= 3)

    let scaledSize = CGSize(width: self.inputSize.width, height: self.inputSize.height)
    guard let thumbnailPixelBuffer = pixelBuffer.centerThumbnail(ofSize: scaledSize) else {
      print("Error: could not crop image")
      return []
    }

    let interval: TimeInterval
    let outputTensor: Tensor
    do {
      let inputTensor = try interpreter.input(at: 0)

      // Remove the alpha component from the image buffer to get the RGB data.
      guard let rgbData = rgbDataFromBuffer(
        thumbnailPixelBuffer,
        isModelQuantized: isModelQuantized
      ) else {
        print("Failed to convert the image buffer to RGB data.")
        return []
      }

      // Copy the RGB data to the input `Tensor`.
      try interpreter.copy(rgbData, toInputAt: 0)

      // Run inference by invoking the `Interpreter`.
      let startDate = Date()
      try interpreter.invoke()
```

cc @jdduke who I believe wrote [this commit](https://github.com/tensorflow/tensorflow/commit/05ed82f0f4c00982cd274a2ebc5d0aeb759eebcb) that seems to be relevant.

**Comment by alexdmiller** [comment]
I tried compiling the select ops framework myself following the [instructions here](https://www.tensorflow.org/lite/guide/ops_select). After including the framework in Xcode, I'm seeing the same error.

Here are screen shots of my Xcode configuration:

<img width="899" alt="Screen Shot 2020-11-20 at 7 01 04 AM" src="https://user-images.githubusercontent.com/926408/99814914-836dc100-2afe-11eb-8f57-821bde391d0d.png">
<img width="739" alt="Screen Shot 2020-11-20 at 7 00 26 AM" src="https://user-images.githubusercontent.com/926408/99814922-85d01b00-2afe-11eb-9a07-61cb4d2f6fcf.png">

**Comment by alexdmiller** [comment]
I noticed on

**Comment by alexdmiller** [comment]
I noticed in the instructions [here](https://www.tensorflow.org/lite/guide/build_ios?hl=fi#modify_xcode_project_settings_directly) that it states: "...you need to add it as an embedded binary to your app target". I tried embedding the `TensorFlowLiteSelectTfOps.framework`, like so:

<img width="747" alt="Screen Shot 2020-11-20 at 9 18 04 AM" src="https://user-images.githubusercontent.com/926408/99829866-b79ead00-2b11-11eb-88dc-227b6277b45b.png">

I then get an error about a missing `Info.plist` when running the app. I added a dummy `Info.plist` to `TensorFlowLiteSelectTfOps.framework`, which makes this error go away. However, the original error ("Regular TensorFlow ops are not supported by this interpreter. Make sure you apply/link the Flex delegate before inference.") still happens when I try to run inference.

**Comment by alexdmiller** [comment]
@teijeong Any suggestions for how to resolve this issue? Thanks!

**Comment by alexdmiller** [comment]
Update: [This section of the Select operators documentation](https://www.tensorflow.org/lite/guide/ops_select#c) suggests that you can build the tensorflow lite library yourself with the select ops included. I attempted to build tensorflow lite myself and include my custom build using the instructions [here](https://www.tensorflow.org/lite/guide/build_ios). However, I am struggling with these instructions; I believe they are incorrect or incomplete. I filed a separate issue [here](https://github.com/tensorflow/tensorflow/issues/45113).

**Comment by alexdmiller** [comment]
@MarkDaoust you were helpful solving [my other problem with the documentation](https://github.com/tensorflow/tensorflow/issues/45113). We haven't heard anything on this issue yet, maybe you have some insight? Thanks!

**Comment by MarkDaoust** [comment]
@yyoon, you did a lot of work on that doc, do you know how to fix this?

**Comment by yyoon** [comment]
Hi @alexdmiller, is there any chance that you're experiencing that behavior only in release mode?
I wonder if this is the same issue as #44891.
cc/ @thaink

**Comment by alexdmiller** [comment]
@yyoon thanks for the suggestion.

I am not very familiar with Xcode and I'm not sure how to tell what mode I'm running the app in. I tried the various options under the "play" button, but I get the same error:

<img width="221" alt="Screen Shot 2020-11-25 at 12 06 00 PM" src="https://user-images.githubusercontent.com/926408/100276618-ad681e80-2f16-11eb-8d30-056c3595eef3.png">

Also, I should add that I made sure to include the linker flags for both "debug" and "release" under build settings.

**Comment by alexdmiller** [comment]
I noticed that the documentation for using CocoaPods has changed slightly ([link](https://www.tensorflow.org/lite/guide/ops_select#using_cocoapods)). It now says to add two flags to Xcode's `Other Linker Flag`s:

```
-force_load $(SRCROOT)/Pods/TensorFlowLiteSelectTfOps/Frameworks/TensorFlowLiteSelectTfOps.framework/TensorFlowLiteSelectTfOps
-u _TF_AcquireFlexDelegate
```

I changed my Podfile to use the nightly TensorFlow build:

```
pod 'TensorFlowLiteSwift', '~> 0.0.1-nightly'
pod 'TensorFlowLiteSelectTfOps', '~> 0.0.1-nightly'
```

Now I'm seeing this error:

```
Undefined symbols for architecture x86_64:
  "_TF_AcquireFlexDelegate", referenced from:
     -u command line option
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by alexdmiller** [comment]
Okay, I just noticed something else in the build logs I hadn't noticed before:

```
Ignoring file Pods/TensorFlowLiteSelectTfOps/Frameworks/TensorFlowLiteSelectTfOps.framework/TensorFlowLiteSelectTfOps, missing required architecture x86_64 in file Pods/TensorFlowLiteSelectTfOps/Frameworks/TensorFlowLiteSelectTfOps.framework/TensorFlowLiteSelectTfOps (2 slices)
```

This definitely seems to be the culprit.

I'm attempting to run my app in the Simulator on my Macbook — is TensorFlowLiteSelectTfOps only built for the phone itself?

**Comment by yyoon** [comment]
Yes, unfortunately we had to remove the `x86_64` architecture support from TensorFlowLiteSelectTfOps framework, because of an issue with CocoaPods: https://github.com/CocoaPods/CocoaPods/issues/8786.
Sorry for this information not being reflected correctly on the official documentation. Will update the doc soon.

If you need to test it with your x86_64 simulator, you would need to [build the framework yourself](https://www.tensorflow.org/lite/guide/ops_select#using_bazel_xcode). The following build command should work:
```
bazel build -c opt --config=ios_x86_64 \
  //tensorflow/lite/ios:TensorFlowLiteSelectTfOps_framework
```

**Comment by Caldarie** [comment]
Hi @alexdmiller,

were you you able to resolve this flex error for iOS? I could get the model working under android, but it seems to be a no go for iOS

**Comment by alexdmiller** [comment]
Thanks!

I was able to get unblocked by running things on an iOS device.

I tried building the framework myself, but I ran into other issues. However, I think those are separate problems and we can close this issue. An update to documentation would be great! Thanks again!

**Comment by yyoon** [comment]
Thanks. FYI, I've already added a note in the doc saying that the prebuilt Select TF ops for iOS doesn't work on x86_64 simulator.
https://www.tensorflow.org/lite/guide/ops_select#ios

**Comment by alexdmiller** [comment]
@yyoon Sorry to revive this issue, but I'm running into it again.

Even though my app works correctly when I build it within Xcode, if I build the app using `xcodebuild` on the command line, I'm seeing the same error:

```
Regular TensorFlow ops are not supported by this interpreter. Make sure you apply/link the Flex delegate before inference.
Node number 312 (FlexSize) failed to prepare.
```

Here is the `xcodebuild` command I'm using:

```
$ xcodebuild -workspace ./MyApp.xcworkspace -scheme MyApp -destination 'generic/platform=iOS' archive -archivePath test-build
```

By running the following command, I double checked that the compiler flags I set within Xcode are being respected by `xcodebuild`. (Note that I've redacted much of this command, but the important part is that we're seeing both the `-force-load` and the `-u _TF_AcquireFlexDelegate` flags.)

```
$ xcodebuild -showBuildSettings -workspace ./MyApp.xcworkspace -scheme MyApp
...
OTHER_LDFLAGS = -force_load Pods/TensorFlowLiteSelectTfOps/Frameworks/TensorFlowLiteSelectTfOps.framework/TensorFlowLiteSelectTfOps -u _TF_AcquireFlexDelegate
...
```

I'm not sure what else could cause the `xcodebuild` app not to load the select op delegates. Any tips would be appreciated.

**Comment by alexdmiller** [comment]
I found [this document](https://gist.github.com/mtauraso/95a49eeb8e5b4335dd2b) that states:

> ACTION: build, test, archive
> 
> What these do is set up in "Edit Scheme" inside of Xcode.app. Archive is special in that it uses stricter header + library search paths. Sometimes the library search paths are sloppy because a sub-project has SKIP_INSTALL set to "false". When "SKIP_INSTALL" is false, libraries are copied during the build action. This copying doesn't happen on archive action resulting in a failure to find the library.

If you notice in the `xcodebuild` command I ran above, I am doing an archive build. Sure enough, the archive build generated from Xcode also fails in the same way (with the `Regular TensorFlow ops are not supported by this interpreter.` error).

So it seems that building an archive does not include the select ops.

I'll continue to investigate but I'm not sure how to fix this. Again, any help appreciated! Thank you!

**Comment by yyoon** [comment]
Possibly related: https://github.com/tensorflow/tensorflow/issues/44879.
Which exact version of TFLite and Select Tf Ops framework are you currently using?

cc: @thaink

**Comment by alexdmiller** [comment]
#44879 seems directly related, that is my exact situation (using TestFlight). I will look at that thread tomorrow morning and see if the fix works for me.

**Comment by alexdmiller** [comment]
@yyoon I followed the instructions in #44879. What ended up ultimately working for me was (a) setting "Strip Linked Product" in Xcode to "No" and (b) updating to the latest version of the library.
